### DISCUSSION TITLE
UEFI: Mellanox ConnectX3 stuck iPXE initializing devices...

### DISCUSSION THREAD

**Discussion by lukts30** [ipxe/ipxe#525] [General]
# UEFI: Mellanox ConnectX3 stuck iPXE initializing devices...

Using: https://aur.archlinux.org/packages/ipxe-git @ https://github.com/ipxe/ipxe/commit/9062544f6a0c69c249b90d21a08d05518aafc2ec 
Same results with @ https://github.com/ipxe/ipxe/commit/e8393c3728bf7073d033410373ef6781549c7c3e

I am using ipxe-hybrid.iso for testing. 
I am trying to get past the "iPXE initializing devices" message. So far the following things were successful:

- BIOS Boot bare-metal (CSM).
- BIOS Boot in qemu/kvm with seabios and PCIe pass-through 
   - with `<timer name="hypervclock" present="yes"/>` set
- UEFI Boot in qemu/kvm with ovmf-edk2 and PCIe pass-through  
   - with `<timer name="hypervclock" present="yes"/>` set
   - capable of sanbooting (iSCSI) a UEFI Arch Linux install with Grub2 as bootloader without any problems (VM is diskless)

The following does not work:

- UEFI Boot bare-metal 
   - iPXE initializing devices...
   - PC resets after 5 minutes 
- UEFI Boot in qemu/kvm with ovmf-edk2 and PCIe pass-through  
   - with `<timer name="hypervclock" present="no"/>` set
   - stuck at: iPXE initializing devices...
- BIOS Boot in qemu/kvm with seabios and PCIe pass-through 
   - with `<timer name="hypervclock" present="no"/>` set
   - stuck at: iPXE initializing devices... 
   - "_" cursor blinks indefinitely 

I only noticed that `hypervclock` "fixes" the issue in the VM because the Windows VM preset just worked while the Linux VM preset got the same problem as if I had booted from bare metal with UEFI. 

I also tried disabling SMT and HPET in UEFI settings but nothing changed though I still can see the HPET clock source as available in Linux.

Without the Mellanox PCIe card installed iPXE never gets stuck and works perfectly fine.
Using the latest iPXE release (v1.21.1) that did not include support for this ConnectX3 card iPXE does also work though of course then I can not make use of the card.

For a VM setup, adding `hypervclock`  is good enough though, **I would still like to get UEFI bare-metal booting working.**

---

System Info:
- X570 AORUS PRO with the latest Firmware (F35d)
- AMD Ryzen 7 5800X
- VM Host: Arch Linux 

Additionally, I get the same problem with the following system. With UEFI bare-metal the system also resets after 5 minutes.
- Asrock X399 Taichi
- AMD Threadripper 1950x
- VM Host: Proxmox 7

---

Additional info:
Linux VM guests always report that TSC is unstable and then switch to kvm-clock but the host OS runs with TSC as `current_clocksource` without problems.

---

Booting of a VM with PCIe pass-through and with and without `hypervclock`:

https://user-images.githubusercontent.com/24390575/143596589-37eaf920-0864-4b5d-83c4-8251dfbcabfb.mp4

**Comment by lukts30** [comment]
Tested a build with: `DEBUG=init:3,rbg,drbg,device,pci,hermon`. Please let me know if there are better Debug options.


Does not work: `<timer name="hypervclock" present="no"/>`:
The same log is also logged when I boot bare-metal with UEFI.

https://user-images.githubusercontent.com/24390575/143613134-810c5639-356d-4dc7-9422-aba2d448cc2e.mp4

The VM consumes 100% CPU on one core.

---

OK: `<timer name="hypervclock" present="yes"/>`:

https://user-images.githubusercontent.com/24390575/143613137-2c860544-9354-41a6-a2ca-521975aca366.mp4

**Comment by lukts30** [comment]
The root cause of this issue is definitely caused by an unreliable clock source. 

https://github.com/ipxe/ipxe/blob/1192edf394aaebde666c6aafda1c8b782b08a845/src/drivers/infiniband/hermon.c#L3794-L3808

The default clock for a bare-metal BIOS (CMS) boot just works on my AMD system and the loop was exited as expected after 500ms. 

With UEFI the code never exited and thous the loop turned into an endless one causing 100% CPU usage. In the VM case using a `hypervclock` fixed this most likely by providing a reliable clock source. 

I have added some debug printf and now see that `currticks()` always returns 0 and therefore:
`(currticks() - start ) ) < HERMON_SENSE_PORT_TIMEOUT )`
`(0-0) < 512` is always true and the loop will run forever.

With hypervclock `currticks()` does return sensible values.

---

For now, I just set the port link-type from VPI to ETH and the problematic loop is never run (exits after the first iteration).

```
sudo mstconfig -d 05:00.0 s LINK_TYPE_P1=ETH
sudo mstconfig -d 05:00.0 s LINK_TYPE_P2=ETH
```

**Comment by lukts30** [comment]
So efi_tick_startup is always "successful" and logs :
>EFI timer started at 32 ticks per second

In a VM without a hypervclock, I can then see that efi_tick gets executed as expected. I simply added the following to efi_tick: 
`DBGC( colour, "called efi_tick %ld\n",efi_jiffies); `

If I now add the Connectx3 PCIe device to VM and try again I see that efi_tick will never be executed the same happens if I boot UEFI bare metal.

**Comment by mcb30** [comment]
@lukts30 Thanks for the detailed debugging work.

There are two immediately plausible explanations for the fact that `efi_tick()` stops being executed:
* Something (possibly iPXE itself) has managed to get confused about the EFI TPL and is preventing the system from dropping back down to `TPL_APPLICATION` (which is the level at which timer interrupts are enabled)
* Something has broken the VM's timer interrupt mechanism.  This is a known issue (with no known root cause) on AMD systems: see e.g. commit https://github.com/ipxe/ipxe/commit/f3ba0fb5f

Given that your issue occurs on bare metal as well as the VM, I'm suspecting the former.  You could try adding debug in `efi_currticks()` to check the value of `efi_external_tpl`, to see if it it somehow never being allowed to drop down to `TPL_APPLICATION`.
